### PR TITLE
[Test] PHPUNIT 6.5 to PHPUNIT 8 | php-ast 0.1.6 to php-ast 1.0.1|

### PR DESCRIPTION
**Pull request by kongtiaowang**
Update PHPUNIT Version
Get rid of dbunit. This package is abandoned and no longer maintained.
The @expectedException, @expectedExceptionCode, @expectedExceptionMessage, and @expectedExceptionMessageRegExp annotations are deprecated. 
They will be removed in PHPUnit 9.  

To test: Travis

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -14,8 +14,7 @@
     },
     "require-dev" : {
         "squizlabs/php_codesniffer" : "2.9.2",
-        "phpunit/phpunit" : "6.5.*",
-        "phpunit/dbunit": "3.0.*",
+        "phpunit/phpunit" : "^8",
```
Comment (driusan) [inline composer.json]: 1. What about the composer.lock file?
2. Are we not using the dbunit package at all? How come we can remove it without any code changes?

Comment (kongtiaowang) [reply composer.json]: We are not using the dbunit, we are using the database mock instead.

**Comment by ridz1208** [comment]
@driusan @kongtiaowang whats the status of this PR ?

**Comment by driusan** [comment]
@ridz1208 it's still missing composer.lock

**Comment by ridz1208** [comment]
@kongtiaowang please make necessary changes

File: composer.json (hunk 1 above)
Comment (kongtiaowang) [reply composer.json]: dbunit is abandoned and no longer maintained officially.

**Comment by johnsaigle** [comment]
@kongtiaowang What's needed to test this PR?

File: composer.json [hunk 2]
```diff
@@ -15,8 +15,7 @@
...
         "facebook/webdriver" : "dev-master",
         "phpmd/phpmd": "^2.6",
         "phan/phan": ">1.2.2"
```
Comment (johnsaigle) [inline composer.json]:
```suggestion
        "phan/phan": ">2.2.3"
```

Since we are updating `php-ast` in this PR, I think it makes sense to update phan to the latest version also

File: .phan/config.php [hunk 3]
```diff
@@ -45,6 +45,7 @@
 	"exclude_analysis_directory_list" => [
 		"vendor",
 		"htdocs/api/",
+                "test/unittests"
```
Comment (johnsaigle) [inline .phan/config.php]: Why are you excluding the unit tests?

**Comment by driusan** [comment]
This still has conflicts and is still excluding the unit tests from phan.

File: .phan/config.php (hunk 3 above)
Comment (kongtiaowang) [reply .phan/config.php]: It will check the PHPUnit library, cause namespace issue.

Comment (kongtiaowang) [reply .phan/config.php]: test/unittests/SinglePointLoginTest.php:44 PhanAccessMethodInternal Cannot access internal method \PHPUnit\Framework\MockObject\Builder\InvocationMocker::will() of namespace \PHPUnit\Framework\MockObject\Builder defined at vendor/phpunit/phpunit/src/Framework/MockObject/Builder/InvocationMocker.php:63 from namespace \

Comment (johnsaigle) [reply .phan/config.php]: I think we should solve the issue instead of excluding all the units tests from phan analysis. Phan can help us detect errors in the unit test code.

In case this is a specific issue for PHP Unit, I think we should try enabling the PHP Unit plugins for phan. https://github.com/phan/phan/tree/master/.phan/plugins#phpunitassertionpluginphp

I can easily add it to this PR (https://github.com/aces/Loris/pull/4964) or make a new one.

Comment (kongtiaowang) [reply .phan/config.php]: @driusan @johnsaigle any ideas? I can't figure it out this namespace issue.

Comment (johnsaigle) [reply .phan/config.php]: I'll make a PR to enable the plugins and we'll see if it helps

File: .phan/config.php [hunk 4]
```diff
@@ -45,6 +45,7 @@
         "vendor/psr",   // fix PhanUndeclaredClass for psr 
         "vendor/zendframework", //fix PhanUndeclaredClass zend 
         "vendor/guzzlehttp",  // fix PhanUndeclaredClass guzzlehttp
+        "vendor/smarty", // fix PhanUndeclaredClass Smarty
```
Comment (johnsaigle) [inline .phan/config.php]: @kongtiaowang I think you can just do `vendor/` and it should get all the subfolders

Comment (kongtiaowang) [reply .phan/config.php]: @johnsaigle It will take more time to do the phan check. I can do it.

File: aaa [hunk 5]
```diff
@@ -0,0 +1 @@
+a
```
Comment (johnsaigle) [inline aaa]: What are these files for?

File: modules/survey_accounts/js/columnFormatter.js [hunk 6]
```diff
@@ -0,0 +1,2 @@
...e)return value;var ns=Object.create(null);__webpack_require__.r(ns);Object.defineProperty(ns,"default",{enumerable:true,value});if(mode&2&&typeof value!="string")for(var key in value)__webpack_require__.d(ns,key,function(key){return value[key]}.bind(null,key));return ns};__webpack_require__.n=function(module){var getter=module&&module.__esModule?function getDefault(){return module["default"]}:function getModuleExports(){return module};__webpack_require__.d(getter,"a",getter);return getter};__webpack_require__.o=function(object,property){return Object.prototype.hasOwnProperty.call(object,property)};__webpack_require__.p="";return __webpack_require__(__webpack_require__.s=63)})({63:function(module,exports,__webpack_require__){"use strict";Object.defineProperty(exports,"__esModule",{value:true});function formatColumn(column,cell,rowData,rowHeaders){var row={};rowHeaders.forEach(function(header,index){row[header]=rowData[index]},this);if(column==="URL"){var url=loris.BaseURL+"/survey.php?key="+row.URL;return React.createElement("td",null,React.createElement("a",{href:url},cell))}return React.createElement("td",null,cell)}window.formatColumn=formatColumn;exports.default=formatColumn}});
```
Comment (johnsaigle) [inline modules/survey_accounts/js/columnFormatter.js]: This file should be removed

File: packages.txt [hunk 7]
```diff
@@ -0,0 +1,56 @@
+ii  libapache2-mod-php5                     5.5.9+dfsg-1ubuntu4.29                                   amd64        server-side, HTML-embedded scripting language (Apache 2 module)
```
Comment (johnsaigle) [inline packages.txt]: What is this for?

File: src/Http/Response/Unauthorized.php [hunk 8]
```diff
@@ -16,7 +16,7 @@
  */
 namespace LORIS\Http\Response;
 
-use \LORIS\Http\Response\JsonResponse;
+use \Zend\Diactoros\Response\JsonResponse;
```
Comment (johnsaigle) [inline src/Http/Response/Unauthorized.php]: I think was correct before. Were you getting an error with this?

**Comment by johnsaigle** [comment]
The .MAP and .BLOB files should be removed also

File: aaa (hunk 5 above)
Comment (kongtiaowang) [reply aaa]: Ops

File: composer.json [hunk 9]
```diff
@@ -15,17 +15,17 @@
...
-        "phan/phan": ">1.2.2"
+        "phan/phan": ">2.2.4"
     },
     "scripts": {
       "pre-install-cmd": "mkdir -p project/libraries"
     },
     "autoload" : {
         "psr-4": {
-            "LORIS\\": "src/"
+            "LORIS\\":  "src/"
```
Comment (PapillonMcGill) [inline composer.json]: extra space, not show blocker
```suggestion
            "LORIS\\": "src/"
```

File: .phan/config.php [hunk 10]
```diff
@@ -23,7 +23,8 @@
 		"PhanTypeMismatchArgument",
 		"PhanTypeMismatchReturn",
 		"PhanTypeMismatchProperty",
-        "PhanTypeSuspiciousStringExpression"
+        "PhanTypeSuspiciousStringExpression",
+        "PhanAccessMethodInternal",
```
Comment (johnsaigle) [inline .phan/config.php]: Why is this added?

File: composer.json [hunk 11]
```diff
@@ -15,10 +15,10 @@
     },
     "require-dev" : {
         "squizlabs/php_codesniffer" : "2.9.2",
-        "phpunit/phpunit" : "6.5.*",
+        "phpunit/phpunit" : "^8",
         "facebook/webdriver" : "dev-master",
         "phpmd/phpmd": "^2.6",
-        "phan/phan": ">1.2.2"
+        "phan/phan": ">2.2.4"
```
Comment (johnsaigle) [inline composer.json]:
```suggestion
        "phan/phan": ">2.2.11"
```

2.2.12 is the latest version as of now.

File: .phan/config.php (hunk 10 above)
Comment (kongtiaowang) [reply .phan/config.php]: fixing the phan error inside the PHPUnit 8 core library.

File: modules/configuration/ajax/updateProject.php [hunk 12]
```diff
@@ -67,8 +67,11 @@
...
 // tables referencing project_subproject_rel in the database.
-$toAdd    = array_diff($_POST['SubprojectIDs'], $preValues);
-$toRemove = array_diff($preValues, $_POST['SubprojectIDs']);
+
+if (isset($_POST['SubprojectIDs'])) {
+    $toAdd    = array_diff($_POST['SubprojectIDs'], $preValues);
+    $toRemove = array_diff($preValues, $_POST['SubprojectIDs']);
+}
 
 foreach ($toAdd as $sid) {
```
Comment (maltheism) [inline modules/configuration/ajax/updateProject.php]: Should line 76 to 94 be moved inside the `if (isset($_POST['SubprojectIDs'])) {..}` or otherwise the $toAdd & $toRemove could possibly be null if isset returns false.

**Comment by kongtiaowang** [comment]
blocked by phan. It should be fixed on other PR.

**Comment by johnsaigle** [comment]
Needs PHPCS on `FILE: /home/travis/build/aces/Loris/php/libraries/Utility.class.inc`

**Comment by johnsaigle** [comment]
Is this still blocked?

**Comment by kongtiaowang** [comment]
It still blocked for no reason.

**Comment by kongtiaowang** [comment]
phpunit > 8.0  is not stable in Travis. failed without reason.
